compiler/rustc_mir_transform/src/dump_mir.rs

```diff
@@ -20,8 +20,15 @@ impl<'tcx> MirPass<'tcx> for Marker {
 }
 
 pub fn emit_mir(tcx: TyCtxt<'_>) -> io::Result<()> {
-    let path = tcx.output_filenames(()).path(OutputType::Mir);
-    let mut f = io::BufWriter::new(File::create(&path)?);
-    write_mir_pretty(tcx, None, &mut f)?;
+    match tcx.output_filenames(()).path(OutputType::Mir) {
+        OutFileName::Stdout => {
+            let mut f = io::stdout();
```

Comment:
I believe we shouldn't override this here and always use `io::BufWriter::new(File::create(&path)?);`

Reply:
Does that mean MIR should always be written to a file? Could you please elaborate on this?

Reply:
MIR is usually always written to `.mir` (`-Zdump-mir`) or graphviz (`-Zdump-mir-graphviz` if I'm not mistaken) files. I don't think `-o -` should override these options, but could be wrong. What seems more consistent with your proposal?

Reply:
Ahhh I got it. So this function (`emit_mir`) is only [called](https://github.com/rust-lang/rust/blob/master/compiler/rustc_interface/src/passes.rs#L877) when the command line contains `--emit mir` or `--emit mir=foo`. I need to override the logic here, because my changes make the commands
```
rustc --emit mir=- test.rs
rustc -o - --emit mir test.rs
```
emit MIR on the terminal, in the same way as
```
rustc -Z unpretty=mir test.rs
```
Otherwise the first two commands above would emit MIR to a file named `-`. As far as I can see, `-Zdump-mir` is handled separately and is not affected by `-o` or `--emit`.


This is how it works if I mix `-Zdump-mir` and `--emit mir=-` together on my end:
```
$ ./build/x86_64-apple-darwin/stage1/bin/rustc -Z dump-mir=all --emit mir=- test.rs
// WARNING: This output format is intended for human consumers only
// and is subject to change without notice. Knock yourself out.
fn main() -> () {
    let mut _0: ();                      // return place in scope 0 at test.rs:1:11: 1:11

    bb0: {
        return;                          // scope 0 at test.rs:2:2: 2:2
    }
}
$ ls -ld mir_dump/
drwxr-xr-x  110 pjhades  staff  3520  1 Jun 10:40 mir_dump/
```
So we have both the stdout output and the MIR dump directory.

Mixing `-o -` and `-Zdump-mir` would not work because the default output type is `link`, which is binary and can't be emitted to a tty:
```
$ ./build/x86_64-apple-darwin/stage1/bin/rustc -Z dump-mir=all -o - test.rs
error: option `-o` or `--emit` is used to write binary output type `link` to stdout, but stdout is a tty

error: aborting due to previous error

$ ls -ld mir_dump/
drwxr-xr-x  110 pjhades  staff  3520  1 Jun 10:44 mir_dump/
```